src/main/java/ir/sahab/kafkaconsumer/SmartCommitKafkaConsumer.java

```diff
@@ -0,0 +1,408 @@
+            @Override
+            public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
+                if (!partitions.isEmpty()) {
+                    logger.warn("Kafka consumer previous assignment revoked: {}", partitions);
+                }
+            }
+
+            @Override
+            public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
+                logger.info("Kafka consumer partitions assigned: {}", partitions);
+                offsetTracker.reset();
+            }
+        };
+        this.unappliedAcks = new ArrayBlockingQueue<>(MAX_UNAPPLIED_ACKS);
+    }
+
+    /**
+     * Subscribes to the given topic by this consumer. Partitions will be assigned dynamically.
+     * Either this method or {@link #assign(String, List)} should be
+     * called just once after construction and before calling {@link #start()}.
+     */
+    public void subscribe(String topic) {
+        checkState(this.topic == null, "It is already subscribed/assigned. You should call either "
+                + "subscribe() or assign() just once.");
+        checkArgument(!Strings.isNullOrEmpty(topic));
+
+        kafkaConsumer.subscribe(Collections.singleton(topic), rebalanceListener);
+        this.topic = topic;
+    }
+
+    /**
+     * Manually assigns the given partition form the requested topic to this consumer.
+     * Either this method or {@link #subscribe(String)} should be
+     * called just once after construction and before calling {@link #start()}.
+     */
+    public void assign(String topic, List<Integer> partitions) {
+        checkState(this.topic == null, "It is already subscribed/assigned. You should call either "
+                + "subscribe() or assign() just once.");
+        checkArgument(!Strings.isNullOrEmpty(topic));
+        requireNonNull(partitions);
+        checkArgument(!partitions.isEmpty());
+
+        List<TopicPartition> topicPartitions = new ArrayList<>();
+        partitions.forEach(p -> topicPartitions.add(new TopicPartition(topic, p)));
+        kafkaConsumer.assign(topicPartitions);
+        this.topic = topic;
+    }
+
+    /**
+     * First ensures connectivity to the Kafka topic and then starts polling from the Kafka topic
+     * in a background thread. Then the polled records can be accessed via {@link #poll()}.
+     * @throws IOException if can not establish connection to the Kafka topic.
+     */
+    public void start() throws InterruptedException, IOException {
+        checkState(topic != null, "You should first call either subscribe() or assign().");
+        checkState(!thread.isAlive(), "start() is called but Kafka consumer is already started.");
+        logger.info("Starting transporter source...");
+
+        // Ensure connection to Kafka topic.
+        ExecutorService executor = Executors.newSingleThreadExecutor();
+        try {
+            executor.submit(() -> kafkaConsumer.poll(0)).get(60, SECONDS);
```

Comment:
We have a problem with Kafka. It blocks forever if it can not connect to the topic. For example we write a wrong topic name, then we see that our application is not working. We see no error log. We see no crash. We just see it is not working. 
So we want to throw Exception if we can not connect at first. (Just as JDBC, just as HBase for example)
This is my workaround solution. I have a future and a timeout, and if I can not return from poll(), I tell the caller that I can not connect. Note that setting timeout as poll() argument does not work, because in Kafka consumer, the section of code which is responsible for connection, is blocking.  It has a TODO in code. ...